## Лабораторная работа№3

Цель работы: практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

Вариант 12

**Задачи лабораторной работы:**
1. Решить свой вариант диофантова уравнения.
2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

In [55]:
from sympy.solvers.diophantine import diophantine
from sympy import symbols
from sympy.crypto.crypto import decipher_rsa
from sympy.crypto.crypto import rsa_private_key
from sympy.logic.boolalg import to_cnf, to_dnf
from sympy.abc import A, B, C
from sympy.logic.boolalg import Xor

**1. Диофантово уравнение:**

Решить диофантово уравнение: 3783x− 2431y = −117

In [56]:
a = 3783
b = 2431
c = 117

x, y = symbols("x, y")
diophantine(a*x - b*y - c)

{(187*t_0 + 81, 291*t_0 + 126)}

81, 126

**2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово:**

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 11,m = 39).
Шифрованное сообщение имеет вид (36; 21; 22; 37). Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “e” и “ъ”).

In [57]:
e = 11
m = 39
message = {36, 21, 22, 37} 
prk = rsa_private_key(m, e)
answer = []
for i in message:
    answer.append(decipher_rsa(i, prk))
print(answer)

[18, 30, 19, 16]


In [58]:
alphabet =['а', 'б', 'в', 'г', 'д', 'е', 'ж', 'з', 'и', 'й', 'к', 'л', 'м', 'н', 'о', 'п',
           'р', 'с', 'т', 'у', 'ф', 'х', 'ц', 'ч', 'ш', 'щ', 'ы', 'ь', 'э', 'ю', 'я']
for i in answer:
    print(alphabet[i - 2])

р
э
с
о


**3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением:**

f(x, y, z) = (x ∨ (z xor y)) ∨ ((yx) xor (xz)):

In [59]:
print("ДНФ:", to_dnf((A | Xor(C, B)) | (Xor((B & A), (A & C))), simplify=True))
print("КНФ:", to_cnf((A | Xor(C, B)) | (Xor((B & A), (A & C))), simplify=True))

ДНФ: A | (B & ~C) | (C & ~B)
КНФ: (A | B | C) & (A | ~B | ~C)


**Вывод:**

В ходе выполнения лабораторной работы была применена на практике библиотека языка Python SymPy, с помощью которой решались задачи по дискретной математике и математической логике.